cpp/test/mdarray.cu

```diff
@@ -436,6 +436,430 @@ TEST(MDArray, FuncArg)
+  auto s             = rmm::cuda_stream_default;
+
+  {
+    constexpr int rows            = 6;
+    constexpr int cols            = 7;
+    constexpr int alignment       = 5;
+    constexpr int alignment_bytes = sizeof(int) * alignment;
+
+    /**
+     * padded device array
+     */
+    using padded_layout_row_major =
+      detail::layout_padded_general<detail::padding<float, alignment_bytes>::value,
+                                    StorageOrderType::row_major_t>;
+
+    using padded_mdarray_type = device_mdarray<float, extents_type, padded_layout_row_major>;
+    padded_layout_row_major::mapping<extents_type> layout(extents_type(rows, cols));
+
+    auto device_policy = padded_mdarray_type::container_policy_type{s};
+    static_assert(std::is_same_v<typename decltype(device_policy)::accessor_type,
+                                 detail::device_uvector_policy<float>>);
+    padded_mdarray_type padded_device_array{layout, device_policy};
+
+    // direct access mdarray
+    padded_device_array(0, 3) = 1;
+    ASSERT_EQ(padded_device_array(0, 3), 1);
+
+    // non-const access via mdspan
+    auto d_view = padded_device_array.view();
+    static_assert(!decltype(d_view)::accessor_type::is_host_type::value);
+
+    thrust::device_vector<int32_t> status(1, 0);
+    auto p_status = status.data().get();
+    thrust::for_each_n(rmm::exec_policy(s),
+                       thrust::make_counting_iterator(0ul),
+                       1,
+                       [d_view, p_status] __device__(size_t i) {
+                         if (d_view(0, 3) != 1) { myAtomicAdd(p_status, 1); }
+                         d_view(0, 2) = 3;
+                         if (d_view(0, 2) != 3) { myAtomicAdd(p_status, 1); }
+                       });
+    check_status(p_status, s);
+
+    // const ref access via mdspan
+    auto const& arr = padded_device_array;
+    ASSERT_EQ(arr(0, 3), 1);
+    auto const_d_view = arr.view();
+    thrust::for_each_n(rmm::exec_policy(s),
+                       thrust::make_counting_iterator(0ul),
+                       1,
+                       [const_d_view, p_status] __device__(size_t i) {
+                         if (const_d_view(0, 3) != 1) { myAtomicAdd(p_status, 1); }
+                       });
+    check_status(p_status, s);
+
+    // initialize with sequence
+    thrust::for_each_n(
+      rmm::exec_policy(s),
+      thrust::make_counting_iterator(0ul),
+      rows * cols,
+      [d_view, rows, cols] __device__(size_t i) { d_view(i / cols, i % cols) = i; });
+
+    // manually create span with layout
+    {
+      auto data_padded         = padded_device_array.data_handle();
+      using padded_mdspan_type = device_mdspan<float, extents_type, padded_layout_row_major>;
+      auto padded_span         = padded_mdspan_type(data_padded, layout);
+      thrust::for_each_n(rmm::exec_policy(s),
+                         thrust::make_counting_iterator(0ul),
```

Comment:
Does Thrust have a way to do `cartesian_product(iota(0, rows), iota(0, cols))` iteration?  It's painful to have to go back and forth between 1-D and 2-D indices like this.

Reply:
I could not find anything like that. Of course, one could define iterators for each dimension and zip them together, but that would just shift the stride-logic to another piece of manual code.